src/libraries/System.Net.Security/src/System/Net/Security/MD4.cs

```diff
@@ -0,0 +1,230 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers.Binary;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Numerics;
+
+//
+// This class is a port of the Mono managed implementation of the MD4 algorithm
+// and required to support NTLM in Android only.
+// It's an implementation detail and is not intended to be a public API.
+// Assuming that NTLM would be System.Net.Security, it makes sense to put MD4 here as well.
+//
+namespace System.Net.Security
+{
+    internal sealed class MD4
+    {
+        private const int S11 = 3;
+        private const int S12 = 7;
+        private const int S13 = 11;
+        private const int S14 = 19;
+        private const int S21 = 3;
+        private const int S22 = 5;
+        private const int S23 = 9;
+        private const int S24 = 13;
+        private const int S31 = 3;
+        private const int S32 = 9;
+        private const int S33 = 11;
+        private const int S34 = 15;
+
+        internal static void HashData(ReadOnlySpan<byte> source, Span<byte> destination)
+        {
+            Debug.Assert(destination.Length == 128 >> 3);
+            Span<byte> buffer = stackalloc byte[64];
+            Span<uint> state = stackalloc uint[4];
+            Span<uint> count = stackalloc uint[2];
+
+            // the initialize our context
+            count[0] = 0;
+            count[1] = 0;
+            state[0] = 0x67452301;
+            state[1] = 0xefcdab89;
+            state[2] = 0x98badcfe;
+            state[3] = 0x10325476;
+            // Zeroize sensitive information
+            buffer.Fill(0);
+
+            HashCore(source, state, count, buffer);
+
+            /* Save number of bits */
+            Span<byte> bits = stackalloc byte[8];
+            Encode(bits, count);
+
+            /* Pad out to 56 mod 64. */
+            uint index = ((count[0] >> 3) & 0x3f);
+            int padLen = (int)((index < 56) ? (56 - index) : (120 - index));
+            Span<byte> padding = stackalloc byte[padLen];
+            padding[0] = 0x80;
+            HashCore(padding, state, count, buffer);
+
+            /* Append length (before padding) */
+            HashCore(bits, state, count, buffer);
+
+            /* Write state to destination */
+            Encode(destination, state);
+        }
+
+        private static void HashCore(ReadOnlySpan<byte> input, Span<uint> state, Span<uint> count, Span<byte> buffer)
+        {
+            /* Compute number of bytes mod 64 */
+            int index = (int)((count[0] >> 3) & 0x3F);
+            /* Update number of bits */
+            count[0] += (uint)(input.Length << 3);
```

Comment:
`input.Length << 3` might overflow the 32-bit space (536870912 bytes or greater).  Presumably this is beyond the design limit of what we want MD4 for, but it's probably worth either handling or just putting an assert in HashData (`input.Length < 0x20000000`).

Reply:
Nevermind, it gets handled on line 77.

::whistles innocently::